fuel-core/src/database/coin.rs

```diff
@@ -83,4 +84,36 @@ impl Database {
             })
         })
     }
+
+    // TODO: Optimize this by creating an index
+    pub fn owned_coins_by_color(
+        &self,
+        owner: Address,
+        color: Color,
+        start_coin: Option<UtxoId>,
+        direction: Option<IterDirection>,
+    ) -> impl Iterator<Item = Result<UtxoId, Error>> + '_ {
+        self.iter_all::<Vec<u8>, bool>(
+            OWNED_COINS,
+            Some(owner.as_ref().to_vec()),
+            start_coin.map(|b| owner_coin_id_key(&owner, &b)),
+            direction,
+        )
+        // Safety: key is always 64 bytes
+        .map(|res| {
+            res.map(|(key, _)| {
+                UtxoId::new(
+                    unsafe { Bytes32::from_slice_unchecked(&key[32..64]) },
+                    key[64],
+                )
+            })
+        })
+        .filter_ok(move |id| {
+            Storage::<UtxoId, Coin>::get(self, id)
+                .unwrap()
```

Comment:
These unwraps are going to be ok in practice currently, but should probably propagate an error in case something else changes the assumptions being made here in the future.

e.g. something like this (untested)

```
.filter_map(|res| {
    match res {
        Ok(id) => {
            Storage::<UtxoId, Coin>::get(self, id)
                  .and_then(|maybe_coin| maybe_coin.ok_or(NotFoundError))
                  .and_then(|coin| if coin.color == color { Some(coin) } else { None })
                  .transpose()
        },
        e => Some(e)
    }
})
```